### PR TITLE
[JENKINS-35246] Kubernetes agents not getting deleted in Jenkins after pods are deleted

### PR DESCRIPTION
**Pull request by Vlatombe**
This fixes the provisioning lifecycle for `KubernetesSlave`. The problem
was `ProvisioningCallback` was adding the node to Jenkins but it is
expected to be done by `NodeProvisioner` once the future completes.

This could lead in cases of short node usage in cases where the node would
get readded to Jenkins after the termination had occurred.

Moving the provisioning logic to a `ComputerLauncher` honors the lifecycle
properly since `ComputerLauncher#launch` is called after the
`NodeProvisioner` has finished its job, eliminating risk of race
condition.

https://issues.jenkins-ci.org/browse/JENKINS-35246

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -212,6 +212,9 @@
             at io.fabric8.kubernetes.client.utils.Serialization.<clinit>(Serialization.java:37)
           -->
           <pluginFirstClassLoader>true</pluginFirstClassLoader>
+          <systemProperties>
+            <hudson.slaves.NodeProvisioner.initialDelay>0</hudson.slaves.NodeProvisioner.initialDelay>
```
Comment (marvinthepa) [inline pom.xml]: 👍

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesSlave.java [hunk 2]
```diff
@@ -96,8 +101,14 @@ public String getNamespace() {
         return namespace;
     }
 
-    public Cloud getCloud() {
-        return Jenkins.getInstance().getCloud(getCloudName());
+    @Nonnull
+    public KubernetesCloud getCloud() {
```
Comment (marvinthepa) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesSlave.java]: I have to admit that I don't understand enough about this, but could this be a binary compatibility issue?

Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesSlave.java]: I restored the original method.

**Comment by marvinthepa** [comment]
@Vlatombe is it possible that the changes from this pull request lead to the fact that a slave can now connect to the master before all containers in that pod have successfully started?

I have noticed this behavior when testing another PR, and this one looks a bit suspicious in this regard.. I have to admit that I didn't check closely, though.

**Comment by Vlatombe** [comment]
@marvinthepa It's possible since the launch is now asynchronous. Setting `computer.setAcceptingTasks(false);` at the beginning of the launcher and `computer.setAcceptingTasks(true);` at the end should do the trick, since the logic already checks that all containers within the pod are launched.

**Comment by dpyoung2** [comment]
Hello,
Will this fix be included in the next release of the plugin? If yes, when is the next release?

**Pull request merged**